nibabel/elemcont.py

```diff
@@ -0,0 +1,188 @@
+    is returned. To get the element itself use the `get_elem` method.
+    '''
+
+    def __init__(self, *args, **kwargs):
+        if len(args) > 1:
+            raise TypeError("At most one arg expected, got %d" % len(args))
+        self._elems = OrderedDict()
+        if len(args) == 1:
+            arg = args[0]
+            if hasattr(arg, 'get_elem'):
+                it = ((k, arg.get_elem(k)) for k in arg)
+            elif hasattr(arg, 'items'):
+                it = iteritems(arg)
+            else:
+                it = arg
+            for key, val in it:
+                self[key] = val
+        for key, val in iteritems(kwargs):
+            self[key] = val
+
+    def __getitem__(self, key):
+        return self._elems[key].value
+
+    def __setitem__(self, key, val):
+        if not hasattr(val, 'value'):
+            raise InvalidElemError(val)
+        self._elems[key] = val
+
+    def __delitem__(self, key):
+        del self._elems[key]
+
+    def __iter__(self):
+        return iter(self._elems)
+
+    def __len__(self):
+        return len(self._elems)
+
+    def __repr__(self):
+        return ('ElemDict(%s)' %
+                ', '.join(['%r=%r' % x for x in self.items()]))
+
+    def update(self, other):
+        if hasattr(other, 'get_elem'):
+            for key in other:
+                self[key] = other.get_elem(key)
+        else:
+            for key, elem in iteritems(other):
+                self[key] = elem
+
+    def get_elem(self, key):
+        return self._elems[key]
+
+
+class ElemList(MutableSequence):
+    '''A list-like container where each value is an "element", which is
+    defined as any object which has a `value` attribute.
+
+    When looking up an item in the list, it is this `value` attribute that
+    is returned. To get the element itself use the `get_elem` method.
+    '''
+    def __init__(self, data=None):
+        self._elems = list()
+        if data is not None:
+            if isinstance(data, self.__class__):
+                for idx in range(len(data)):
+                    self.append(data.get_elem(idx))
+            else:
+                for elem in data:
+                    self.append(elem)
+
+    def _tuple_from_slice(self, slc):
+        '''Get (start, end, step) tuple from slice object.
+        '''
+        (start, end, step) = slc.indices(len(self))
+        # Replace (0, -1, 1) with (0, 0, 1) (misfeature in .indices()).
+        if step == 1:
+          if end < start:
+            end = start
+          step = None
+        if slc.step == None:
+          step = None
+        return (start, end, step)
+
+    def __getitem__(self, idx):
+        if isinstance(idx, slice):
+            return ElemList(self._elems[idx])
+        else:
+            return self._elems[idx].value
+
+    def __setitem__(self, idx, val):
+        if isinstance(idx, slice):
+            (start, end, step) = self._tuple_from_slice(idx)
+            if step != None:
```

Comment:
Maybe (indentation again):

```
if step is None:
    # Normal slice
    for j, assign_val in enumerate(val):
        self.insert(start + j, assign_val)
        return
# Extended slice
...
```

Sorry to keep on with this kind of thing, but I personally find that easier to read, because then, after I've read some `if something: do something; return` clause, I know I can forget about that clause.  Otherwise I've got to look further down the method to see whether something else is happening as well.